### PR TITLE
[#175679733] Add GetSupportToken API

### PR DESCRIPTION
**Pull request by AleDore**
This PR adds a new endpoint `/support-token` to io-backend API's in order to provide a JWT token for a logged user, who wants to open an Instabug's issue by IO app.

### CODE REVIEW

File: api_backend.yaml
Comment (gunzip) [inline api_backend.yaml:1047]: what's this value? where's the start date saved? can it be expires_at with a date instead?

Comment (BurnedMarshal) [reply api_backend.yaml:1047]: Commonly the use of `expire_in` is preferred because doesn't need clock synchronization to verify (client-side) when the token expires.

Comment (gunzip) [reply api_backend.yaml:1047]: if the access_token is a JWT, doesn't it contains the `exp` claim itself?

File: api_backend.yaml
Comment (gunzip) [inline api_backend.yaml:1050]: can we have different token types?

File: api_backend.yaml
Comment (AleDore) [reply api_backend.yaml:1047]: Yes it contains `exp` claim, that contains the Date (in seconds) of token expiration, while `expires_in` contains a time expression of token validity. For example if we generate a support token right now and the TTL for this token is `7 days` we have:
- `exp` contains the date expressed in seconds for next week;
- `expires_in` contains `604800` (the number of seconds in 7 days)

File: api_backend.yaml
Comment (AleDore) [reply api_backend.yaml:1050]: Nope, for this moment :)

File: api_backend.yaml
Comment (BurnedMarshal) [reply api_backend.yaml:1047]: Yes, it does. The `exp` claim is a timestamp and could be extracted with token introspection. At server level clock synchronization is required and we can use `clockTolerance` for configuring the clock screw. `expire_in` add more information only on client side.

File: api_backend.yaml
Comment (gunzip) [reply api_backend.yaml:1050]: why we need to return this value then?

Comment (AleDore) [reply api_backend.yaml:1050]: we base our implementation on this standard:
```
{
    "access_token": "{your access token}",
    "expires_in": {15 minutes expressed as seconds},
    "token_type": "Bearer"
}
```

File: api_backend.yaml
Comment (BurnedMarshal) [inline api_backend.yaml:480]: Maybe we can use a generic url instead `support-token`, for example `/token`
Can we use the existing `AuthenticationController`?

File: api_backend.yaml
Comment (gunzip) [reply api_backend.yaml:1050]: why?

**Review by balanza — COMMENTED** [review]
Apart from minor things, the code looks good to me.

I have some concerns about api design.
I'm sure this has already been analysed deeply, still I feel like we are patching an endpoint to satisfy an emergent need quickly. Don't we already have an IDP mechanism for sharing the user's session with a third party service?

I think we should spend more time motivating the _why_s that led to this implementation. In my opinion that's very important to avoid design quality to degrade steadily at every business need.

File: api_backend.yaml
Comment (BurnedMarshal) [reply api_backend.yaml:1050]: The proposed interface refers to the Auth0 solution. We only use Bearer token, for simplicity we can omit this parameter.

File: api_backend.yaml
Comment (AleDore) [reply api_backend.yaml]: Endpoint changed :). At the moment, I'd rather keep `SupportController` for other needs if necessary :)

File: api_backend.yaml
Comment (gunzip) [reply api_backend.yaml:1050]: Yes, as a general guidelines a parameter must be _motivated_ and _documented_ otherwise we are going to accumulate tech debt on things that we don't remember the raison d'etre of.

**Comment by AleDore** [comment]
> Apart from minor things, the code looks good to me.
> 
> I have some concerns about api design.
> I'm sure this has already been analysed deeply, still I feel like we are patching an endpoint to satisfy an emergent need quickly. Don't we already have an IDP mechanism for sharing the user's session with a third party service?
> 
> I think we should spend more time motivating the _WHYs_ that led to this implementation. In my opinion that's very important to avoid design quality to degrade steadily at every business need.

It's true, but in this case `supportToken` is not used as an authentication token, but it is used as a param sent through instabug, that allows support flow to reach user's issue data without using fiscalCode (for security reasons).

**Comment by balanza** [comment]
> It's true, but in this case `supportToken` is not used as an authentication token, but it is used as a param sent through instabug, that allows support flow to reach user's issue data without using fiscalCode (for security reasons).

I'm super sure the technical solution is good, and of course I have full trust in everyone how commit here. 
My point is: a PR like this comes up and there's no clue on how can we discuss about api design, if it's consistent with previous implementations and if it's aligned with the product roadmap.
But maybe is a broader subject and we can discuss it in another place.

**Comment by AleDore** [comment]
> > It's true, but in this case `supportToken` is not used as an authentication token, but it is used as a param sent through instabug, that allows support flow to reach user's issue data without using fiscalCode (for security reasons).
> 
> I'm super sure the technical solution is good, and of course I have full trust in everyone how commit here.
> 
> My point is: a PR like this comes up and there's no clue on how can we discuss about api design, if it's consistent with previous implementations and if it's aligned with the product roadmap.
> But maybe is a broader subject and we can discuss it in another place.

Maybe we have to discuss our API design roadmap into a dedicated RFC :) don't we ?

File: src/config.ts
Comment (gunzip) [inline src/config.ts:388]:
```suggestion
  "JWT_SUPPORT_TOKEN_PRIVATE_RSA_KEY"
```

File: src/config.ts
Comment (gunzip) [inline src/config.ts:398]:
```suggestion
  "JWT_SUPPORT_TOKEN_ISSUER"
```

File: .env.example
Comment (gunzip) [inline .env.example:45]: I guess we cannot merge & deploy this PR since these values are required until they're set up in the  production app service, isn't it?

**Review by gunzip — CHANGES_REQUESTED** [review]

File: .env.example
Comment (AleDore) [reply .env.example:45]: Yes I'll prepare another PR on infrastructure :)

Comment (AleDore) [reply .env.example:45]: Done [https://github.com/pagopa/io-infrastructure-live-new/pull/290](https://github.com/pagopa/io-infrastructure-live-new/pull/290)

Comment (gunzip) [reply .env.example:45]: lgtm, once applied we can merge this and deploy

**Review by gunzip — APPROVED** [review]

**Comment by gunzip** [comment]
@AleDore @BurnedMarshal is this ready to be deployed with all envs set up?

File: src/config.ts
Comment (gunzip) [inline src/config.ts:409]: can you try to reafactor using `IntegerFromString.decode(process.env.JWT_SUPPORT_TOKEN_EXPIRATIO)` ?

**Review by gunzip — CHANGES_REQUESTED** [review]

File: src/config.ts
Comment (AleDore) [reply src/config.ts:409]: Yep

**Review by gunzip — APPROVED** [review]

File: api_backend.yaml
Comment (Undermaken) [inline api_backend.yaml:479]:
```suggestion
  "/token/support":
```
or
```suggestion
  "/support_token":
```
Perhaps we would add other tokens in the future.
Could it be a better name?

**Review by Undermaken — CHANGES_REQUESTED** [review]

File: api_backend.yaml
Comment (AleDore) [reply api_backend.yaml:479]: Initially it was `"/support-token` https://github.com/pagopa/io-backend/pull/726#discussion_r522191997, but we have changed it in order to mantain a generic approach, but you're right cc: @BurnedMarshal @gunzip

**Pull request merged**
